File: src/main/java/tool/arenasizer/ArenaPanel.java
package tool.arenasizer;

import core.gui.comp.entry.ColorLabelEntry;
import core.gui.comp.entry.DoubleLabelEntries;
import core.gui.comp.entry.IHOTableCellEntry;
import core.gui.comp.renderer.HODefaultTableCellRenderer;
import core.model.HOModel;
import core.model.HOVerwaltung;
import core.model.TranslationFacility;
import core.util.Helper;
import tool.updater.TableModel;

import javax.swing.*;
import javax.swing.table.TableColumnModel;
import java.awt.*;
import java.math.BigDecimal;
import java.util.Optional;
import java.util.stream.Stream;

final class ArenaPanel extends JPanel {

    private static final String[] ROW_TRANSLATION_KEYS = {"ls.club.arena.terraces", "ls.club.arena.basicseating", "ls.club.arena.seatsunderroof", "ls.club.arena.seatsinvipboxes", "Gesamt", "Einnahmen", "Unterhalt", "Gewinn", "Baukosten"};

    private static final int ROW_INDEX_TERRACES = 0;
    private static final int ROW_INDEX_BASIC_SEATING = 1;
    private static final int ROW_INDEX_SEATS_UNDER_ROOF = 2;
    private static final int ROW_INDEX_VIP = 3;
    private static final int ROW_INDEX_SUM = 4;
    private static final int ROW_INDEX_REVENUE = 5;
    private static final int ROW_INDEX_WEEKLY_COST = 6;
    private static final int ROW_INDEX_PROFIT = 7;
    private static final int ROW_INDEX_CONSTRUCTION_COST = 8;
    private static final int NUMBER_OF_ROWS = 9;

    private static final String[] COLUMN_TRANSLATION_KEYS_TITLES = {null, "Aktuell", "Maximal", "Durchschnitt", "Minimal"};

    private static final int COLUMN_INDEX_TITLE = 0;
    private static final int COLUMN_INDEX_ACTUAL = 1;
    private static final int COLUMN_INDEX_MAXIMUM = 2;
    private static final int COLUMN_INDEX_AVERAGE = 3;
    private static final int COLUMN_INDEX_MINIMUM = 4;
    private static final int NUMBER_OF_COLUMNS = 5;

    private final JTable jTable = new JTable();

    private Stadium currentStadium;
    private Stadium[] stadiumsMaxAvgMin;
    private IHOTableCellEntry[][] tableCellEntries;

    ArenaPanel() {
        setLayout(new BorderLayout());
        add(new JScrollPane(jTable));
        jTable.setDefaultRenderer(Object.class, new HODefaultTableCellRenderer());
        jTable.getTableHeader().setReorderingAllowed(false);
        initTable();
        reInit();
    }

    public void reInit() {
        HOModel model = HOVerwaltung.instance().getModel();
        currentStadium = model.getStadium();
        stadiumsMaxAvgMin = ArenaSizer.calcConstructionArenas(currentStadium, model.getClub().getFans());
        reinitTable();
    }

    private void initTable() {
        // set table values
        tableCellEntries = new IHOTableCellEntry[NUMBER_OF_ROWS][NUMBER_OF_COLUMNS];

        for (int rowIndex = 0; rowIndex < ROW_TRANSLATION_KEYS.length; rowIndex++) {
            tableCellEntries[rowIndex][COLUMN_INDEX_TITLE] = new ColorLabelEntry(TranslationFacility.tr(ROW_TRANSLATION_KEYS[rowIndex]), ColorLabelEntry.FG_STANDARD, ColorLabelEntry.BG_PLAYERSPOSITIONVALUES, SwingConstants.LEFT);
        }

        // Placeholders
        for (int row = ROW_INDEX_TERRACES; row < NUMBER_OF_ROWS; row++) {
            for (int column = COLUMN_INDEX_ACTUAL; column < NUMBER_OF_COLUMNS; column++) {
                if (row < COLUMN_INDEX_MINIMUM) {
                    tableCellEntries[row][column] = createDoubleLabelEntries(ColorLabelEntry.BG_PLAYERSSUBPOSITIONVALUES);
                } else if (row == COLUMN_INDEX_MINIMUM) {
                    tableCellEntries[row][column] = createDoubleLabelEntries(ColorLabelEntry.BG_PLAYERSPOSITIONVALUES);
                } else {
                    tableCellEntries[row][column] = createDoubleLabelEntries(ColorLabelEntry.BG_SINGLEPLAYERVALUES);
                }
            }
        }

        jTable.setModel(new TableModel(tableCellEntries, createColumnTitles()));

        final TableColumnModel columnModel = jTable.getColumnModel();
        columnModel.getColumn(COLUMN_INDEX_TITLE).setMinWidth(Helper.calcCellWidth(150));
        columnModel.getColumn(COLUMN_INDEX_ACTUAL).setMinWidth(Helper.calcCellWidth(160));
        columnModel.getColumn(COLUMN_INDEX_MAXIMUM).setMinWidth(Helper.calcCellWidth(160));
        columnModel.getColumn(COLUMN_INDEX_AVERAGE).setMinWidth(Helper.calcCellWidth(160));
        columnModel.getColumn(COLUMN_INDEX_MINIMUM).setMinWidth(Helper.calcCellWidth(160));
    }

    /**
     * Create a new DoubleLabelEntries with default values
     *
     * @param background Color
     * @return DoubleLabelEntries
     */
    private static DoubleLabelEntries createDoubleLabelEntries(Color background) {
        return new DoubleLabelEntries(new ColorLabelEntry("", ColorLabelEntry.FG_STANDARD, background, SwingConstants.RIGHT), new ColorLabelEntry("", ColorLabelEntry.FG_STANDARD, background, SwingConstants.RIGHT));
    }

    void reinitArena(Stadium currentArena, int maxSupporter, int normalSupporter, int minSupporter) {
        currentStadium = currentArena;
        stadiumsMaxAvgMin = ArenaSizer.calcConstructionArenas(currentArena, maxSupporter, normalSupporter, minSupporter);
        reinitTable();
    }

    private void reinitTable() {
        final Stadium stadium = HOVerwaltung.instance().getModel().getStadium();
        if (currentStadium != null) {
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_TERRACES][COLUMN_INDEX_ACTUAL]).getLeft().setText(formatNumber(currentStadium.getTerraces()));
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_TERRACES][COLUMN_INDEX_ACTUAL]).getRight().setSpecialNumber(currentStadium.getTerraces() - stadium.getTerraces(), false);
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_BASIC_SEATING][COLUMN_INDEX_ACTUAL]).getLeft().setText(formatNumber(currentStadium.getBasicSeating()));
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_BASIC_SEATING][COLUMN_INDEX_ACTUAL]).getRight().setSpecialNumber(currentStadium.getBasicSeating() - stadium.getBasicSeating(), false);
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_SEATS_UNDER_ROOF][COLUMN_INDEX_ACTUAL]).getLeft().setText(formatNumber(currentStadium.getUnderRoofSeating()));
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_SEATS_UNDER_ROOF][COLUMN_INDEX_ACTUAL]).getRight().setSpecialNumber(currentStadium.getUnderRoofSeating() - stadium.getUnderRoofSeating(), false);
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_VIP][COLUMN_INDEX_ACTUAL]).getLeft().setText(formatNumber(currentStadium.getVipBox()));
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_VIP][COLUMN_INDEX_ACTUAL]).getRight().setSpecialNumber(currentStadium.getVipBox() - stadium.getVipBox(), false);
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_SUM][COLUMN_INDEX_ACTUAL]).getLeft().setText(formatNumber(currentStadium.getTotalSize()));
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_SUM][COLUMN_INDEX_ACTUAL]).getRight().setSpecialNumber(currentStadium.getTotalSize() - stadium.getTotalSize(), false);
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_REVENUE][COLUMN_INDEX_ACTUAL]).getLeft().setText(ArenaAdmission.calculateIncome(currentStadium).toLocaleString());
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_REVENUE][COLUMN_INDEX_ACTUAL]).getRight().setText(ArenaAdmission.calculateIncome(currentStadium).minus(ArenaAdmission.calculateIncome(stadium)).toLocaleString());
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_WEEKLY_COST][COLUMN_INDEX_ACTUAL]).getLeft().setText(ArenaMaintenance.calculateCosts(currentStadium).times(BigDecimal.valueOf(-1)).toLocaleString());
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_WEEKLY_COST][COLUMN_INDEX_ACTUAL]).getRight().setText(ArenaMaintenance.calculateCosts(currentStadium).minus(ArenaMaintenance.calculateCosts(stadium)).times(BigDecimal.valueOf(-1)).toLocaleString());
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_PROFIT][COLUMN_INDEX_ACTUAL]).getLeft().setText(ArenaAdmission.calculateIncome(currentStadium).minus(ArenaMaintenance.calculateCosts(currentStadium)).toLocaleString());
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_PROFIT][COLUMN_INDEX_ACTUAL]).getRight().setText(ArenaAdmission.calculateIncome(currentStadium).minus(ArenaMaintenance.calculateCosts(currentStadium)).minus(ArenaAdmission.calculateIncome(stadium)).minus(ArenaMaintenance.calculateCosts(stadium)).toLocaleString());
            final int newTerraces = currentStadium.getTerraces() - stadium.getTerraces();
            final int newBasicSeating = currentStadium.getBasicSeating() - stadium.getBasicSeating();
            final int newUnderRoofSeating = currentStadium.getUnderRoofSeating() - stadium.getUnderRoofSeating();
            final int newVipBox = currentStadium.getVipBox() - stadium.getVipBox();
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_CONSTRUCTION_COST][COLUMN_INDEX_ACTUAL]).getLeft().setText(ArenaRebuild.calculateCosts(newTerraces, newBasicSeating, newUnderRoofSeating, newVipBox).toLocaleString());
            ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_CONSTRUCTION_COST][COLUMN_INDEX_ACTUAL]).getRight().setText("");

            final int startColumn = COLUMN_INDEX_MAXIMUM;
            for (int column = startColumn; column < NUMBER_OF_COLUMNS; column++) {
                final int stadiumTypeIndex = column - startColumn;
                final Stadium stadiumType = stadiumsMaxAvgMin[stadiumTypeIndex];
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_TERRACES][column]).getLeft().setText(formatNumber(stadiumType.getTerraces()));
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_TERRACES][column]).getRight().setSpecialNumber(stadiumType.getTerraces() - currentStadium.getTerraces(), false);
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_BASIC_SEATING][column]).getLeft().setText(formatNumber(stadiumType.getBasicSeating()));
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_BASIC_SEATING][column]).getRight().setSpecialNumber(stadiumType.getBasicSeating() - currentStadium.getBasicSeating(), false);
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_SEATS_UNDER_ROOF][column]).getLeft().setText(formatNumber(stadiumType.getUnderRoofSeating()));
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_SEATS_UNDER_ROOF][column]).getRight().setSpecialNumber(stadiumType.getUnderRoofSeating() - currentStadium.getUnderRoofSeating(), false);
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_VIP][column]).getLeft().setText(formatNumber(stadiumType.getVipBox()));
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_VIP][column]).getRight().setSpecialNumber(stadiumType.getVipBox() - currentStadium.getVipBox(), false);
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_SUM][column]).getLeft().setText(formatNumber(stadiumType.getTotalSize()));
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_SUM][column]).getRight().setSpecialNumber(stadiumType.getTotalSize() - currentStadium.getTotalSize(), false);
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_REVENUE][column]).getLeft().setText(ArenaAdmission.calculateIncome(stadiumType).toLocaleString());
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_REVENUE][column]).getRight().setText(ArenaAdmission.calculateIncome(stadiumType).minus(ArenaAdmission.calculateIncome(currentStadium)).toLocaleString());
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_WEEKLY_COST][column]).getLeft().setText(ArenaMaintenance.calculateCosts(stadiumType).times(BigDecimal.valueOf(-1)).toLocaleString());
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_WEEKLY_COST][column]).getRight().setText(ArenaMaintenance.calculateCosts(stadiumType).minus(ArenaMaintenance.calculateCosts(currentStadium)).times(BigDecimal.valueOf(-1)).toLocaleString());
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_PROFIT][column]).getLeft().setText(ArenaAdmission.calculateIncome(stadiumType).minus(ArenaMaintenance.calculateCosts(stadiumType)).toLocaleString());
                ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_PROFIT][column]).getRight().setText(ArenaAdmission.calculateIncome(stadiumType).minus(ArenaMaintenance.calculateCosts(stadiumType)).minus(ArenaAdmission.calculateIncome(currentStadium)).minus(ArenaMaintenance.calculateCosts(currentStadium)).toLocaleString());
                var expansionCosts = stadiumType.getExpansionCosts();
                if (expansionCosts != null) {
                    ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_CONSTRUCTION_COST][column]).getLeft().setText(expansionCosts.times(BigDecimal.valueOf(-1)).toLocaleString());
                } else {
                    ((DoubleLabelEntries) tableCellEntries[ROW_INDEX_CONSTRUCTION_COST][column]).getLeft().setText("");
                }
            }

            jTable.setModel(new TableModel(tableCellEntries, createColumnTitles()));
            jTable.getColumnModel().getColumn(COLUMN_INDEX_TITLE).setMinWidth(Helper.calcCellWidth(150));
            jTable.getColumnModel().getColumn(COLUMN_INDEX_ACTUAL).setMinWidth(Helper.calcCellWidth(160));
            jTable.getColumnModel().getColumn(COLUMN_INDEX_MAXIMUM).setMinWidth(Helper.calcCellWidth(160));
            jTable.getColumnModel().getColumn(COLUMN_INDEX_AVERAGE).setMinWidth(Helper.calcCellWidth(160));
            jTable.getColumnModel().getColumn(COLUMN_INDEX_MINIMUM).setMinWidth(Helper.calcCellWidth(160));
        }
    }

    private static Object[] createColumnTitles() {
        return Stream.of(COLUMN_TRANSLATION_KEYS_TITLES).map(key -> Optional.ofNullable(key).map(TranslationFacility::tr).orElse("")).toArray();
    }

    private static String formatNumber(int number) {
        final var numberformat = Helper.getNumberFormat(0);
        return numberformat.format(number);
    }
}

File: src/main/java/tool/arenasizer/ArenaPanel.kt
package tool.arenasizer

import core.gui.comp.entry.ColorLabelEntry
import core.gui.comp.entry.DoubleLabelEntries
import core.gui.comp.entry.IHOTableEntry
import core.gui.comp.renderer.HODefaultTableCellRenderer
import core.model.HOVerwaltung
import core.util.Helper
import tool.updater.TableModel
import java.awt.BorderLayout
import java.awt.Color
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.JTable
import javax.swing.SwingConstants

/**
 * Panel for Stadium display and testing
 */
class ArenaPanel : JPanel() {
    //~ Instance fields ----------------------------------------------------------------------------
    private val arenaSizer = ArenaSizer()
    private val arenaTable = JTable()

    //Teststadium
    private var stadium: Stadium? = null
    private lateinit var stadiumArray: Array<Stadium>
    private lateinit var values: Array<Array<IHOTableEntry?>>

    private val HEADERS = arrayOf(
        "",
        HOVerwaltung.instance().getLanguageString("Aktuell"),
        HOVerwaltung.instance().getLanguageString("Maximal"),
        HOVerwaltung.instance().getLanguageString("Durchschnitt"),
        HOVerwaltung.instance().getLanguageString("Minimal")
    )


    //~ Constructors -------------------------------------------------------------------------------
    init {
        setLayout(BorderLayout())
        add(JScrollPane(arenaTable))
        arenaTable.setDefaultRenderer(Any::class.java, HODefaultTableCellRenderer())
        arenaTable.tableHeader.setReorderingAllowed(false)
        initTable()
        reInit()
    }

    //-------Refresh---------------------------------
    fun reInit() {
        val model = HOVerwaltung.instance().model
        stadium = model.getStadium()
        stadiumArray = arenaSizer.calcConstructionArenas(stadium!!, model.getClub().fans)

        reinitTable()
    }

    private fun initTable() {
        values = Array(9) { arrayOfNulls(5) }
        val hoV = HOVerwaltung.instance()
        val columnText = arrayOf(
            "ls.club.arena.terraces",
            "ls.club.arena.basicseating",
            "ls.club.arena.seatsunderroof",
            "ls.club.arena.seatsinvipboxes",
            "Gesamt",
            "Einnahmen",
            "Unterhalt",
            "Gewinn",
            "Baukosten"
        )

        for (i in columnText.indices) {
            values[i][0] = ColorLabelEntry(
                hoV.getLanguageString(columnText[i]),
                ColorLabelEntry.FG_STANDARD, ColorLabelEntry.BG_PLAYERSPOSITIONVALUES, SwingConstants.LEFT
            )
        }

        //Platzwerte
        for (i in 0..8) {
            for (j in 1..4) {
                if (i < 4) values[i][j] = createDoppelLabelEntry(ColorLabelEntry.BG_PLAYERSSUBPOSITIONVALUES)
                else if (i == 4) values[i][j] = createDoppelLabelEntry(ColorLabelEntry.BG_PLAYERSPOSITIONVALUES)
                else values[i][j] = createDoppelLabelEntry(ColorLabelEntry.BG_SINGLEPLAYERVALUES)
            }
        }
        arenaTable.setModel(TableModel(values, HEADERS))
        val columnModel = arenaTable.columnModel
        columnModel.getColumn(0).setMinWidth(Helper.calcCellWidth(150))
        columnModel.getColumn(1).setMinWidth(Helper.calcCellWidth(160))
        columnModel.getColumn(2).setMinWidth(Helper.calcCellWidth(160))
        columnModel.getColumn(3).setMinWidth(Helper.calcCellWidth(160))
        columnModel.getColumn(4).setMinWidth(Helper.calcCellWidth(160))
    }

    /**
     * create a new DoppelLabelEntry with default values
     * @param background
     * @return
     */
    private fun createDoppelLabelEntry(background: Color): DoubleLabelEntries {
        return DoubleLabelEntries(
            ColorLabelEntry(
                "",
                ColorLabelEntry.FG_STANDARD,
                background, SwingConstants.RIGHT
            ),
            ColorLabelEntry(
                "",
                ColorLabelEntry.FG_STANDARD,
                background, SwingConstants.RIGHT
            )
        )
    }

    fun reinitArena(currentArena: Stadium, maxSupporter: Int, normalSupporter: Int, minSupporter: Int) {
        stadium = currentArena
        stadiumArray = arenaSizer.calcConstructionArenas(currentArena, maxSupporter, normalSupporter, minSupporter)

        //Entrys mit Werten füllen
        reinitTable()
    }

    private fun reinitTable() {
        val stadium = HOVerwaltung.instance().model.getStadium()
        val currentStadium = this.stadium!!

        (values[0][1] as DoubleLabelEntries).left.setText("${currentStadium.standing}")
        (values[0][1] as DoubleLabelEntries).right.setSpecialNumber(
            currentStadium.standing - stadium.standing,
            false
        )

        (values[1][1] as DoubleLabelEntries).left.setText("${currentStadium.basicSeating}")
        (values[1][1] as DoubleLabelEntries).right.setSpecialNumber(
            currentStadium.basicSeating - stadium.basicSeating,
            false
        )

        (values[2][1] as DoubleLabelEntries?)!!.left.setText("${currentStadium.seatingUnderRoof}")
        (values[2][1] as DoubleLabelEntries?)!!.right.setSpecialNumber(
            currentStadium.seatingUnderRoof - stadium.seatingUnderRoof,
            false
        )

        (values[3][1] as DoubleLabelEntries?)!!.left.setText("${currentStadium.vip}")
        (values[3][1] as DoubleLabelEntries?)!!.right.setSpecialNumber(
            currentStadium.vip - stadium.vip,
            false
        )

        (values[4][1] as DoubleLabelEntries?)!!.left.setText("${currentStadium.totalSize()}")
        (values[4][1] as DoubleLabelEntries?)!!.right.setSpecialNumber(
            currentStadium.totalSize() - stadium.totalSize(),
            false
        )

        (values[5][1] as DoubleLabelEntries?)!!.left.setSpecialNumber(
            arenaSizer.calcMaxIncome(currentStadium),
            true
        )
        (values[5][1] as DoubleLabelEntries?)!!.right.setSpecialNumber(
            arenaSizer.calcMaxIncome(currentStadium) - arenaSizer.calcMaxIncome(
                stadium
            ), true
        )

        (values[6][1] as DoubleLabelEntries?)!!.left.setSpecialNumber(
            -arenaSizer.calcMaintenance(currentStadium),
            true
        )
        (values[6][1] as DoubleLabelEntries?)!!.right.setSpecialNumber(
            -(arenaSizer.calcMaintenance(currentStadium) - arenaSizer.calcMaintenance(
                stadium
            )), true
        )

        (values[7][1] as DoubleLabelEntries?)!!.left.setSpecialNumber(
            arenaSizer.calcMaxIncome(currentStadium) - arenaSizer.calcMaintenance(currentStadium),
            true
        )
        (values[7][1] as DoubleLabelEntries?)!!.right.setSpecialNumber(
            arenaSizer.calcMaxIncome(currentStadium) - arenaSizer.calcMaintenance(currentStadium) - (arenaSizer.calcMaxIncome(stadium)
                    - arenaSizer.calcMaintenance(stadium)), true
        )

        (values[8][1] as DoubleLabelEntries?)!!.left.setSpecialNumber(
            -arenaSizer.calcConstructionCosts(
                (currentStadium.standing - stadium.standing).toFloat(),
                (currentStadium.basicSeating - stadium.basicSeating).toFloat(),
                (currentStadium.seatingUnderRoof - stadium.seatingUnderRoof).toFloat(),
                (currentStadium.vip - stadium.vip).toFloat()
            ), true
        )
        (values[8][1] as DoubleLabelEntries?)!!.right.setText("")
        for (i in 2..4) {
            (values[0][i] as DoubleLabelEntries?)!!.left.setText(
                stadiumArray[i - 2].standing.toString() + ""
            )
            (values[0][i] as DoubleLabelEntries?)!!.right.setSpecialNumber(
                stadiumArray[i - 2].standing - currentStadium.standing,
                false
            )
            (values[1][i] as DoubleLabelEntries?)!!.left.setText(
                stadiumArray[i - 2].basicSeating.toString() + ""
            )
            (values[1][i] as DoubleLabelEntries?)!!.right.setSpecialNumber(
                stadiumArray[i - 2].basicSeating - currentStadium.basicSeating,
                false
            )
            (values[2][i] as DoubleLabelEntries?)!!.left.setText(
                stadiumArray[i - 2].seatingUnderRoof.toString() + ""
            )
            (values[2][i] as DoubleLabelEntries?)!!.right.setSpecialNumber(
                stadiumArray[i - 2].seatingUnderRoof - currentStadium.seatingUnderRoof,
                false
            )
            (values[3][i] as DoubleLabelEntries?)!!.left.setText(stadiumArray[i - 2].vip.toString() + "")
            (values[3][i] as DoubleLabelEntries?)!!.right.setSpecialNumber(
                stadiumArray[i - 2].vip - currentStadium.vip,
                false
            )
            (values[4][i] as DoubleLabelEntries?)!!.left.setText(
                stadiumArray[i - 2].totalSize().toString() + ""
            )
            (values[4][i] as DoubleLabelEntries?)!!.right.setSpecialNumber(
                stadiumArray[i - 2].totalSize() - currentStadium.totalSize(),
                false
            )
            (values[5][i] as DoubleLabelEntries?)!!.left.setSpecialNumber(
                arenaSizer.calcMaxIncome(
                    stadiumArray[i - 2]
                ), true
            )
            (values[5][i] as DoubleLabelEntries?)!!.right.setSpecialNumber(
                arenaSizer.calcMaxIncome(
                    stadiumArray[i - 2]
                ) - arenaSizer.calcMaxIncome(currentStadium), true
            )
            (values[6][i] as DoubleLabelEntries?)!!.left.setSpecialNumber(
                -arenaSizer.calcMaintenance(
                    stadiumArray[i - 2]
                ), true
            )
            (values[6][i] as DoubleLabelEntries?)!!.right.setSpecialNumber(
                -(arenaSizer.calcMaintenance(
                    stadiumArray[i - 2]
                )
                        - arenaSizer.calcMaintenance(currentStadium)), true
            )
            (values[7][i] as DoubleLabelEntries?)!!.left.setSpecialNumber(
                arenaSizer.calcMaxIncome(
                    stadiumArray[i - 2]
                ) - arenaSizer.calcMaintenance(stadiumArray[i - 2]), true
            )
            (values[7][i] as DoubleLabelEntries?)!!.right.setSpecialNumber(
                arenaSizer.calcMaxIncome(
                    stadiumArray[i - 2]
                ) - arenaSizer.calcMaintenance(stadiumArray[i - 2])
                        - (arenaSizer.calcMaxIncome(currentStadium) - arenaSizer.calcMaintenance(currentStadium)),
                true
            )
            (values[8][i] as DoubleLabelEntries?)!!.left.setSpecialNumber(
                -stadiumArray[i - 2].expansionCosts,
                true
            )
        }

        arenaTable.setModel(TableModel(values, HEADERS))

        arenaTable.columnModel.getColumn(0).setMinWidth(Helper.calcCellWidth(150))
        arenaTable.columnModel.getColumn(1).setMinWidth(Helper.calcCellWidth(160))
        arenaTable.columnModel.getColumn(2).setMinWidth(Helper.calcCellWidth(160))
        arenaTable.columnModel.getColumn(3).setMinWidth(Helper.calcCellWidth(160))
        arenaTable.columnModel.getColumn(4).setMinWidth(Helper.calcCellWidth(160))
    }
}
